### PR TITLE
Avoid loops with local-redirect service translation

### LINKED ISSUE #11646: Cilium egress policies should support local node redirection
Tools like kiam (https://github.com/uswitch/kiam) and kube2iam (https://github.com/jtblin/kube2iam) use iptables rules to redirect traffic destined to the cloud metadata service (169.254.169.254:80) to a local daemonset running in host networking.   These iptables rules match on the source device name (e.g., lxc+ for cilium) in order to ensure they match only pod traffic, and not traffic from the host itself.  

However, if pods an an egress L7 proxy redirection policy to enable HTTP visibility on TCP/80 policy for all destinations, traffic to the metadata service will be redirect to the L7 proxy (which is OK), but then the traffic leaving this proxy and going to the outside world will not match the iptables redirect rule, because the source device will not longer by lxcXXXX.  

One possible long-term path for this would be to implement the type of redirection done by kiam in iptables directly in BPF.   It is a pretty simple DNAT rule: 

```
iptables \
  --append PREROUTING \
  --protocol tcp \
  --destination 169.254.169.254 \
  --dport 80 \
  --in-interface lxc+ \
  --jump DNAT \
  --table nat \
  --to-destination <node-ip>:8181
```

This is likely what we would want to do for enabling kiam/kube2iam in a iptables free env anyway.

Note that we would likely want this rule inserted via a config time option, or via cluster-wide network policy if we wanted a bit more configurability, but not via standard CNP, as we wouldn't want tenants being able to disable this.   It is similar to Node firewall in that sense.

### PR DESCRIPTION
**Pull request by aditighag**
Service translation logic for a `local-redirect` service can cause packets to be looped back to a service node-local backend after translation. This can happen when the node-local pod itself tries to connect to the service frontend for which it acts as a backend. There are cases (e.g., kiam agent connecting the AWS metadata server on EKS) where this can break traffic flow if the backend needs to forward the redirected traffic to the actual service frontend.

Check commit description for more details about the issue and testing.

AWS metadata server doesn't seem to support IPv6 [1], hence the fix is applicable only to IPv4. IPv6 changes can be added in future.

Follow-ups:  end-to-end tests, update gsg with kernel requirement for the socket lookup

Fixes: #11646

[1] https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

### CODE REVIEW

**Comment by aditighag** [comment]
test-me-please

**Comment by aditighag** [comment]
Marking the PR as draft to look into the CI failures.

**Comment by aditighag** [comment]
test-me-please

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -279,6 +279,8 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	struct lb4_service *backend_slot;
 	bool backend_from_affinity = false;
 	__u32 backend_id = 0;
+	struct bpf_sock_tuple tuple = {};
+	struct bpf_sock *sk = NULL;
```
Comment (borkmann) [inline bpf/bpf_sock.c]: This needs to go under scope of `if (lb4_svc_is_localredirect(svc))` body, otherwise we need this memset for all non-LPR traffic in fast path.

File: bpf/bpf_sock.c [hunk 2]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     *
+     * For example, a local-redirect service exists with a frontend
+     * <169.254.169.254, 80> and backend be1. When traffic destined to the
+     * frontend originates from be1 in namespace ns1 (host ns in case of
+     * hostNetwork pods or child ns for regular pods), when be1 is selected as a
+     * backend, the traffic would get looped backed to be1.
+     * Identify such cases by doing a socket lookup for be1 <ip, port> in its
+     * namespace, ns1, and skip service translation.
+     */
+	if (lb4_svc_is_localredirect(svc)) {
```
Comment (borkmann) [inline bpf/bpf_sock.c]: Pls move all of this including comment into a small helper function that is self-descriptive, e.g. `sock4_backend_is_same_netns()`.

Also the `struct bpf_sock_tuple` doesn't need to do memset twice below? I'd just init the struct on declaration.

I'd remove the `For a negative signed netns` comment, and just have the code be self-descriptive. E.g. set a define SAME_NETNS -1, and we should be good.

File: bpf/include/bpf/helpers.h [hunk 3]
```diff
@@ -84,4 +84,15 @@ static int BPF_FUNC(sock_hash_update, struct bpf_sock_ops *skops, void *map,
 static int BPF_FUNC(msg_redirect_hash, struct sk_msg_md *md, void *map,
 		    void *key, __u64 flags);
 
+/* Socket lookup helpers */
+/* sk_lookup_* apply to various contexts, including socket, sk_buff, etc.
+ * Hence, use void pointer type.
+ */
```
Comment (borkmann) [inline bpf/include/bpf/helpers.h]: I'd just remove the comment `apply to various contexts` since I think it's obvious enough from the used code.

File: bpf/lib/common.h [hunk 4]
```diff
@@ -555,7 +555,11 @@ enum {
 	SVC_FLAG_LOADBALANCER = (1 << 5),  /* LoadBalancer service */
 	SVC_FLAG_ROUTABLE     = (1 << 6),  /* Not a surrogate/ClusterIP entry */
 	SVC_FLAG_SOURCE_RANGE = (1 << 7),  /* Check LoadBalancer source range */
-	SVC_FLAG_LOCALREDIRECT = (1 << 8),  /* local redirect */
+};
+
+/* Service flags (lb{4,6}_service->flags2) */
+enum {
+	SVC_FLAG_LOCALREDIRECT = (1 << 0),  /* local redirect */
 };
```
Comment (borkmann) [inline bpf/lib/common.h:568]: So this means the current code in master never worked, or just by accident? Pls split this out into a separate commit with Fixes tag. Also, please work on the flags{,2}->__u16 conversion so this is not needed.

File: bpf/bpf_sock.c (hunk 2 above)
Comment (borkmann) [reply bpf/bpf_sock.c]: Please also add CI tests. Especially also for v4-in-v6 sockets.

**Review by borkmann — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
@aditighag I've changed the release note label since this is not a user facing change and we aren't fixing a bug of a released feature.

File: bpf/lib/common.h (hunk 4 above)
Comment (aditighag) [reply bpf/lib/common.h:568]: The flag wasn't used until now.

>please work on the flags{,2}->__u16 conversion 

The current changes were done based on your suggestion/approval (https://github.com/cilium/cilium/pull/12831#issuecomment-681833150). I'm curious to know why you prefer one over the other?

File: bpf/include/bpf/helpers.h (hunk 3 above)
Comment (aditighag) [reply bpf/include/bpf/helpers.h]: Well, the function is currently not used in any other context, but I'm fine either way.

File: bpf/bpf_sock.c [hunk 5]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+	    memset(&tuple.ipv4.sport, 0, sizeof(tuple.ipv4.sport));
+	    tuple.ipv4.daddr = backend->address;
+	    tuple.ipv4.dport = backend->port;
+
+	    switch (ctx->protocol) {
+            case IPPROTO_TCP:
+                /* For a negative signed netns, the ns associated with the ctx
+                 * (in this case, socket) will be used.
+                 */
+                sk = sk_lookup_tcp(ctx_full, &tuple, sizeof(tuple.ipv4), -1, 0);
```
Comment (joestringer) [inline bpf/bpf_sock.c]:
```suggestion
                sk = sk_lookup_tcp(ctx_full, &tuple, sizeof(tuple.ipv4), BPF_F_CURRENT_NETNS, 0);
```

Comment (joestringer) [reply bpf/bpf_sock.c]: Same below.

File: bpf/include/bpf/helpers.h (hunk 3 above)
Comment (joestringer) [reply bpf/include/bpf/helpers.h]: sk_lookup_* is used for BPF TProxy from skb context.

```
$ git grep sk_lookup
bpf/include/bpf/helpers_skb.h:static struct bpf_sock *BPF_FUNC(sk_lookup_udp, struct __sk_buff *skb,
bpf/include/linux/bpf.h: * struct bpf_sock *bpf_sk_lookup_tcp(void *ctx, struct bpf_sock_tuple *tuple, u32 tuple_size, u64 netns, u64 flags)
bpf/include/linux/bpf.h: * struct bpf_sock *bpf_sk_lookup_udp(void *ctx, struct bpf_sock_tuple *tuple, u32 tuple_size, u64 netns, u64 flags)
bpf/include/linux/bpf.h: *              **bpf_sk_lookup_xxx**\ ().
bpf/include/linux/bpf.h: *              This function is identical to **bpf_sk_lookup_tcp**\ (), except
bpf/include/linux/bpf.h: *              returned from **bpf_sk_lookup_xxx**\ (),
bpf/include/linux/bpf.h:        FN(sk_lookup_tcp),              \
bpf/include/linux/bpf.h:        FN(sk_lookup_udp),              \
bpf/lib/proxy.h:        sk = sk_lookup_udp(ctx, tuple, len, BPF_F_CURRENT_NETNS, 0);
```

**Review by joestringer — CHANGES_REQUESTED** [review]
Is it possible to add tests for this?

File: bpf/bpf_sock.c (hunk 5 above)
Comment (aditighag) [reply bpf/bpf_sock.c]: Ah! I did come across `BPF_F_CURRENT_NETNS`, but I wanted to dig into the code to see how it is updated.

File: bpf/bpf_sock.c [hunk 6]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 		return -ENOENT;
 	}
 
+    /* Service translation logic for a local-redirect service can cause
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: s/local-redirect/local-policy-redirect/

File: bpf/bpf_sock.c [hunk 7]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     * packets to be looped back to a service node-local backend after translation.
+     * This can happen when the node-local backend itself tries to connect to the
+     * service frontend for which it acts as a backend.
+     * There are cases where this can break traffic flow if the backend needs to
+     * forward the redirected traffic to the actual service frontend.
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: I'd mention here an exact example for future readers.

File: bpf/bpf_sock.c [hunk 8]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     * Hence, allow service translation for pod traffic getting redirected to
+     * backend (across namespaces), but skip service translation for backend to
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: s/across namespaces/across network namespaces/ to avoid ambiguity.

File: bpf/bpf_sock.c [hunk 9]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     * to itself or another service backend within the same namespace.
+     *
+     * For example, a local-redirect service exists with a frontend
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: s/local-redirect/local-policy-redirect/

File: bpf/bpf_sock.c [hunk 10]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     * <169.254.169.254, 80> and backend be1. When traffic destined to the
+     * frontend originates from be1 in namespace ns1 (host ns in case of
+     * hostNetwork pods or child ns for regular pods), when be1 is selected as a
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: s/child//

File: bpf/bpf_sock.c [hunk 11]
```diff
@@ -345,6 +347,53 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
...
+     * backend, the traffic would get looped backed to be1.
```
Comment (brb) [inline bpf/bpf_sock.c]: Nit: s/backed/back/

**Review by brb — COMMENTED** [review]
I'm wondering whether there might be a valid case for LRP in which a request from LRP backend is expected to be sent back to the same backend. Such case would be broken, wouldn't it?

File: bpf/bpf_sock.c (hunk 9 above)
Comment (aditighag) [reply bpf/bpf_sock.c]: `local-redirect` _is_ the service type.

File: bpf/bpf_sock.c (hunk 6 above)
Comment (aditighag) [reply bpf/bpf_sock.c]: `local-redirect` _is_ the service type.

**Comment by aditighag** [comment]
Trying to figure out how to test ipv6 functionality.

File: bpf/bpf_sock.c (hunk 6 above)
Comment (brb) [reply bpf/bpf_sock.c]: Aha, I thought that the LRP service type is called `local-redirect-policy`.

**Comment by aditighag** [comment]
test-me-please

File: bpf/include/bpf/helpers_skb.h [hunk 12]
```diff
@@ -54,9 +54,6 @@ static int BPF_FUNC_REMAP(skb_event_output, struct __sk_buff *skb, void *map,
 static struct bpf_sock *BPF_FUNC(skc_lookup_tcp, struct __sk_buff *skb,
 				 struct bpf_sock_tuple *tuple, __u32 tuple_size,
 				 __u64 netns, __u64 flags);
-static struct bpf_sock *BPF_FUNC(sk_lookup_udp, struct __sk_buff *skb,
```
Comment (joestringer) [inline bpf/include/bpf/helpers_skb.h:57]: Why is this helper being deleted from the `helpers_skb.h`?

File: bpf/bpf_sock.c [hunk 13]
```diff
@@ -264,6 +264,61 @@ sock4_wildcard_lookup_full(struct lb4_key *key __maybe_unused,
...
+    /* There exists a socket listening on the <ip, port> of the backend
+     * in the same namespace where socket connect call originates from.
+     */
+     if (sk) {
+         sk_release(sk);
+         return true;
+     }
+
+	return false;
+}
```
Comment (joestringer) [inline bpf/bpf_sock.c:319]: This code should also be wrapped in `#ifdef BPF_HAVE_SOCKET_LOOKUP` shouldn't it?

File: bpf/include/bpf/helpers.h [hunk 14]
```diff
@@ -84,4 +84,12 @@ static int BPF_FUNC(sock_hash_update, struct bpf_sock_ops *skops, void *map,
 static int BPF_FUNC(msg_redirect_hash, struct sk_msg_md *md, void *map,
 		    void *key, __u64 flags);
 
+/* Socket lookup helpers */
+static struct bpf_sock *BPF_FUNC(sk_lookup_tcp, void *ctx,
+				 struct bpf_sock_tuple *tuple, __u32 tuple_size,
+				 __u64 netns, __u64 flags);
+static struct bpf_sock *BPF_FUNC(sk_lookup_udp, void *ctx,
```
Comment (joestringer) [inline bpf/include/bpf/helpers.h:91]: In the skb variants, we are using the actual context type as the second parameter (`struct __sk_buff *`), does it make sense to do that here with `bpf_sock_addr *`? This would allow compile-time type checks on the parameters I think?

**Review by joestringer — CHANGES_REQUESTED** [review]
A few more nits below. I would expect the skbuff change should actually cause CI to fail (though I guess it's a little early to tell).

Checkpatch GH action is also complaining a lot, did you run `make -C bpf checkpatch` before submitting?

File: bpf/bpf_sock.c (hunk 13 above)
Comment (aditighag) [reply bpf/bpf_sock.c:319]: This will be called conditionally since it's already wrapped under `ifdef`? Are you suggesting that we add the ifdefs in case this function is called some place else in future?

File: bpf/include/bpf/helpers_skb.h (hunk 12 above)
Comment (aditighag) [reply bpf/include/bpf/helpers_skb.h:57]: I added them to helpers.h with void *ctx.

File: bpf/include/bpf/helpers.h (hunk 14 above)
Comment (aditighag) [reply bpf/include/bpf/helpers.h:91]: Not sure, this never came up in previous review comments - https://github.com/cilium/cilium/pull/13287#discussion_r494817243

**Review by aditighag — COMMENTED** [review]
> Checkpatch GH action is also complaining a lot, did you run `make -C bpf checkpatch` before submitting?

I ran it just now. It's complaining about indentations at some places. I'll fix it.

**Comment by aditighag** [comment]
Fixed formatting errors flagged by checkpatch.

**Comment by aditighag** [comment]
test-me-please

File: bpf/include/bpf/helpers_skb.h (hunk 12 above)
Comment (joestringer) [reply bpf/include/bpf/helpers_skb.h:57]: Somehow I was misinterpreting the change to be introducing specific helpers for SK rather than sharing the same helper everywhere 🤦. I was also confused by how one of the socket helpers remains here but it's because that one is not used more generally, only in the skb contexts. So all is fine here, sorry for the noise.

File: bpf/include/bpf/helpers.h (hunk 14 above)
Comment (joestringer) [reply bpf/include/bpf/helpers.h:91]: It's not such a big deal, partly the background of this question is my misunderstanding on introducing additional variations of the helper for different contexts rather than just sharing common declarations. I'll mark this as resolved.

File: bpf/bpf_sock.c (hunk 13 above)
Comment (aditighag) [reply bpf/bpf_sock.c:319]: I've wrapped the function as well under `ifdef` in case it's called at some other place in future.

**Comment by joestringer** [comment]
There's still a couple of minor checkpatch warnings, I think kernel style expects tabs to be 8 chars wide:

```
CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
#99: FILE: bpf/bpf_sock.c:289:
+sock4_skip_xlate_for_same_netns_backend(struct bpf_sock_addr *ctx,
+					     const struct lb4_backend *backend)

CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
#141: FILE: bpf/bpf_sock.c:407:
+	if (lb4_svc_is_localredirect(svc) &&
+		sock4_skip_xlate_for_same_netns_backend(ctx_full, backend)) {


NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.
```

Also need to triage the unsuccessful checks.

**Review by brb — COMMENTED** [review]
The datapath changes LGTM. A few things:

- We should log an info msg about consequences in the daemon if sk_lookup_{tcp,udp} is missing and users have enabled LRP.
- Considering that we have a testing infrastructure for LRP, could we add a e2e test for it? From a first glance, it doesn't sound like a lot of work. Just need to create two pods - one accessible via LRP, another via ClusterIP. The former should send a request to the latter via ClusterIP.
- Do you know whether there might be a valid case for LRP in which a request from LRP backend is expected to be sent back to the same backend?

**Comment by aditighag** [comment]
> There's still a couple of minor checkpatch warnings, I think kernel style expects tabs to be 8 chars wide:
> 
> ```
> CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
> #99: FILE: bpf/bpf_sock.c:289:
> +sock4_skip_xlate_for_same_netns_backend(struct bpf_sock_addr *ctx,
> +					     const struct lb4_backend *backend)
> 
> CHECK:PARENTHESIS_ALIGNMENT: Alignment should match open parenthesis
> #141: FILE: bpf/bpf_sock.c:407:
> +	if (lb4_svc_is_localredirect(svc) &&
> +		sock4_skip_xlate_for_same_netns_backend(ctx_full, backend)) {
> 
> 
> NOTE: For some of the reported defects, checkpatch may be able to
>       mechanically convert to the typical style using --fix or --fix-inplace.
> ```
> 
> Also need to triage the unsuccessful checks.

I'm not able to see this in the IDE, and the --fix-inplace isn't working.

**Comment by aditighag** [comment]
> The datapath changes LGTM. A few things:
> 
> * We should log an info msg about consequences in the daemon if sk_lookup_{tcp,udp} is missing and users have enabled LRP.

Do we have an equivalent of `HAVE_PROG_TYPE_HELPER ` from datapath in the cilium agent code? Can you point me to it? 

> * Considering that we have a testing infrastructure for LRP, could we add a e2e test for it? 

Absolutely! It's added in the follow-ups list. Let me see if I can it to this same PR.

> * Do you know whether there might be a valid case for LRP in which a request from LRP backend is expected to be sent back to the same backend?

It seems unlikely. Currently we don't have such a use case.

**Comment by brb** [comment]
> Do we have an equivalent of HAVE_PROG_TYPE_HELPER from datapath in the cilium agent code? Can you point me to it?
@aditighag In the daemon we check for the helpers, e.g. https://github.com/cilium/cilium/blob/master/daemon/cmd/kube_proxy_replacement.go#L150. Perhaps you could do the same for `sk_lookup_{tcp,udp}`.

**Comment by aditighag** [comment]
test-net-next

**Comment by aditighag** [comment]
I wrapped the code path under ifdef of enable-ipv6 so that it's disabled for IPv4 mapped IPv6 mapping. Per the description, IPv6 isn't required at the moment. With this, the kernel net-next is now passing. PTAL.

**Comment by aditighag** [comment]
test-4.19

**Comment by aditighag** [comment]
test-4.9

**Comment by aditighag** [comment]
test-runtime

**Comment by aditighag** [comment]
test-ginkgo

**Comment by aanm** [comment]
retest-gke

**Review by brb — APPROVED** [review]
LGTM, minus the checkpatch failure.

**Comment by aditighag** [comment]
> LGTM, minus the checkpatch failure.

checkpatch is giving a warning about one of the lines in the commit description being longer than the certain characters limit.

**Comment by joestringer** [comment]
> checkpatch is giving a warning about one of the lines in the commit description being longer than the certain characters limit.

:+1: This doesn't make sense to restrict when including wider sample output of commands in the commit message, it's more useful to have it exactly how it is right now rather than to make checkpatch happy in this specific case. Maybe we should consider relaxing checkpatch in this area.

File: bpf/bpf_sock.c [hunk 15]
```diff
@@ -264,6 +264,63 @@ sock4_wildcard_lookup_full(struct lb4_key *key __maybe_unused,
...
+ * For example, in EKS cluster, a local-redirect service exists with the AWS
+ * metadata IP, port as the frontend <169.254.169.254, 80> and kiam proxy as a
+ * backend pod. When traffic destined to the frontend originates from the kiam pod
+ * in namespace ns1 (host ns when the kiam proxy pod is deployed in hostNetwork
+ * mode or regular pod ns) and the pod is selected as a backend, the traffic
+ * would get looped back to the proxy pod.
+ * Identify such cases by doing a socket lookup for the backend <ip, port> in its
+ * namespace, ns1, and skip service translation.
+ */
+#ifndef ENABLE_IPV6
```
Comment (joestringer) [inline bpf/bpf_sock.c]: nit: Any particular reason why `#ifndef ENABLE_IPV6` rather than `#ifdef ENABLE_IPV4`?

File: bpf/bpf_sock.c [hunk 16]
```diff
@@ -347,6 +404,15 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 		return -ENOENT;
 	}
 
+#ifndef ENABLE_IPV6
```
Comment (joestringer) [inline bpf/bpf_sock.c]: Does this mean that if you simply enable IPv6, then it will break the connectivity for IPv4? (or more specifically, reintroduce this loop issue)

Even if the KIAM metadata server may not support IPv6 for some time, I don't think that precludes running IPv6 applications over the network for other purposes.

Wondering if you considered passing down the IPv4 vs. IPv6 context from the caller (as bool for example) so that we just skip this step for the sock6_xlate_v4_in_v6() case?

File: bpf/bpf_sock.c (hunk 15 above)
Comment (aditighag) [reply bpf/bpf_sock.c]: Well, in this particular case, we want to disable this code path for ipv6 context, particularly for ipv4 mapped ipv6 addresses processing.

File: bpf/bpf_sock.c (hunk 16 above)
Comment (aditighag) [reply bpf/bpf_sock.c]: Sure, that sounds like a better solution to me.

**Comment by aditighag** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by borkmann** [comment]
> AWS metadata server doesn't seem to support IPv6 [1], hence the fix is applicable only to IPv4. IPv6 changes can be added in future.

> Per the description, IPv6 isn't required at the moment.

As a user I would expect v4 to behave exactly the same as v6 for this particular service type, even if AWS only implements it in v4 today ... but users might have other use cases as well, so we should be consistent and avoid surprises where we later on would have to break existing behavior. Lets better get it implemented now before it gets used in the wild.

**Comment by borkmann** [comment]
Would also be good to document the semantics wrt same netns in https://docs.cilium.io/en/latest/gettingstarted/local-redirect-policy/ .
